politeiawww/paywall.go

```diff
@@ -1,47 +1,280 @@
+	if err != nil {
+		return err
+	}
+
+	user.NewUserPaywallAddress = address
+	user.NewUserPaywallAmount = amount
+	user.NewUserPaywallTxNotBefore = txNotBefore
+	err = b.db.UserUpdate(*user)
+	if err != nil {
+		return err
+	}
+
+	b.AddUserToPaywallPool(user)
+	return nil
+}
+
+// ProcessVerifyUserPayment verifies that the provided transaction
 // meets the minimum requirements to mark the user as paid, and then does
 // that in the user database.
-func (b *backend) ProcessVerifyUserPaymentTx(user *database.User, vupt www.VerifyUserPaymentTx) (*www.VerifyUserPaymentTxReply, error) {
+func (b *backend) ProcessVerifyUserPayment(user *database.User, vupt v1.VerifyUserPayment) (*v1.VerifyUserPaymentReply, error) {
+	var reply v1.VerifyUserPaymentReply
+	if b.HasUserPaid(user) {
+		reply.HasPaid = true
+		return &reply, nil
+	}
+
+	if paywallHasExpired(user.NewUserPaywallTxNotBefore) {
+		b.GenerateNewUserPaywall(user)
+
+		reply.PaywallAddress = user.NewUserPaywallAddress
+		reply.PaywallAmount = user.NewUserPaywallAmount
+		reply.PaywallTxNotBefore = user.NewUserPaywallTxNotBefore
+		return &reply, nil
+	}
+
 	minConfirmations := b.cfg.MinConfirmationsRequired
-	verified, err := util.VerifyTxWithBlockExplorers(user.NewUserPaywallAddress,
-		user.NewUserPaywallAmount, vupt.TxId, user.NewUserPaywallTxNotBefore, minConfirmations)
+	txId, err := util.FetchTxWithBlockExplorers(user.NewUserPaywallAddress,
+		user.NewUserPaywallAmount, user.NewUserPaywallTxNotBefore, minConfirmations)
 	if err != nil {
+		if err == util.ErrCannotVerifyPayment {
+			return nil, v1.UserError{
+				ErrorCode: v1.ErrorStatusCannotVerifyPayment,
+			}
+		}
 		return nil, err
 	}
 
-	var reply www.VerifyUserPaymentTxReply
-	if verified {
+	if txId != "" {
 		reply.HasPaid = true
 
-		user.NewUserPaywallTx = vupt.TxId
+		user.NewUserPaywallTx = txId
 		err = b.db.UserUpdate(*user)
 		if err != nil {
 			return nil, err
 		}
+	} else {
+		// Add the user to the in-memory pool.
 	}
 
 	return &reply, nil
 }
 
-// VerifyUserPaid checks that a user has paid the paywall
-func (b *backend) VerifyUserPaid(user *database.User) bool {
+// HasUserPaid checks that a user has paid the paywall
+func (b *backend) HasUserPaid(user *database.User) bool {
 	// Return true when running unit tests
 	if b.test {
 		return true
 	}
 
 	// Return true if paywall is disabled
-	if b.cfg.PaywallAmount == 0 && b.cfg.PaywallXpub == "" {
+	if !b.PaywallIsEnabled() {
 		return true
 	}
 
 	return user.NewUserPaywallTx != ""
 }
+
+// AddUserToPaywallPool adds a user and its paywall info to the in-memory pool.
+func (b *backend) AddUserToPaywallPool(user *database.User) {
+	mutex.Lock()
+	addUser(user)
+	mutex.Unlock()
+}
+
+// InitPaywallCheck is intended to be called
+func (b *backend) InitPaywallCheck() error {
+	if b.cfg.PaywallAmount == 0 {
+		// Paywall not configured.
+		return nil
+	}
+
+	// Create the in-memory pool of all users who need to pay the paywall.
+	mutex.Lock()
```

Comment:
Deadlock if there is any failure.

Per last comment, make the bits that need to be locked into a function if possible. Much easier to keep track of locking.

Add doco and mark that the call must be called WITHOUT the mutex held.